File: server/wheelroads/post.php
<?php
	require_once 'auth.php';
	require_once 'connect.php';
	require_once 'get_post.php';
	require_once 'status.php';
	require_once 'success.php';
	error_reporting( E_ALL );

	// 当該低アクセシビリティ道路への情報(口コミ)の投稿
	// postsテーブルへ登録する
	// post_idのみを返す
	// JSON

	$road_id = get( 'road_id' );
	$is_steps = get_bool( 'is_steps' );
	$is_difference = get_bool( 'is_difference' );
	$is_steep = get_bool( 'is_steep' );
	$is_rough = get_bool( 'is_rough' );
	$is_narrow = get_bool( 'is_narrow' );
	$is_cant = get_bool( 'is_cant' );
	$is_bikes = get_bool( 'is_bikes' );
	$level = get( 'level' );
	$comment = get_string( 'comment' );
	$lang = get_string( 'lang' );

	// $image_flag = isset( $_FILE['upfile']['tmp_name'] );
	// echo $_FILE['upfile']['tmp_name'];

	if ( strlen( $lang ) != 2 ) {
		status( "INVALID LENGTH LANG PARAMETER $lang" );
	}

	$lang = strtolower( $lang );

	$user_id = auth();

	try {
		$dbh = connect();

		$dbh->beginTransaction();

		try {
			$sql = "INSERT INTO posts (
								road_id, user_id,
								is_steps, is_difference, is_steep, is_rough,
								is_narrow, is_cant, is_bikes,
								level, comment, lang
							)
			 				VALUES (
								:road_id, :user_id,
								:is_steps, :is_difference, :is_steep, :is_rough,
								:is_narrow, :is_cant, :is_bikes,
								:level, :comment, :lang
							)";

			$stmt = $dbh->prepare( $sql );
			$stmt->bindValue( ':road_id', $road_id, PDO::PARAM_INT );
			$stmt->bindValue( ':user_id', $user_id, PDO::PARAM_INT );
			$stmt->bindValue( ':is_steps', $is_steps, PDO::PARAM_INT );
			$stmt->bindValue( ':is_difference', $is_difference, PDO::PARAM_INT );
			$stmt->bindValue( ':is_steep', $is_steep, PDO::PARAM_INT );
			$stmt->bindValue( ':is_rough', $is_rough, PDO::PARAM_INT );
			$stmt->bindValue( ':is_narrow', $is_narrow, PDO::PARAM_INT );
			$stmt->bindValue( ':is_cant', $is_cant, PDO::PARAM_INT );
			$stmt->bindValue( ':is_bikes', $is_bikes, PDO::PARAM_INT );
			$stmt->bindValue( ':level', $level, PDO::PARAM_INT );
			$stmt->bindParam( ':comment', $comment, PDO::PARAM_STR );
			$stmt->bindParam( ':lang', $lang, PDO::PARAM_STR );

			/*
			if ( $image_flag ) {
				//$uploaddir = '/srv/http/wheelroads/';
				//$uploadfile = $uploaddir.basename( $_FILES['upfile']['name'] );
				$image = file_get_contents( $_FILES['upfile']['tmp_name'] );
				$stmt->bindParam( ':image', $image, PDO::PARAM_STR );
			}
			*/

			$flag = $stmt->execute();
			$post_id = $dbh->lastInsertId();

			if ( !$flag ) {
				//print_r( $stmt->errorInfo() );
				$dbh->rollBack();
				status( 'COULD NOT INSERT POST' );
			}

			$dbh->commit();
		} catch ( Exception $e ) {
			$dbh->rollBack();
			status( 'TRANSACTION ERROR' );
		}

		success( array( 'post_id' => $post_id ) );
	} catch ( PDOException $e ) {
		status( 'DATABASE ERROR' );
	} catch ( Exception $e ) {
		status( 'UNKNOWN ERROR' );
	}
?>

File: app/src/main/java/com/example/foobar/wheelroads/Post.java
package com.example.foobar.wheelroads;

import android.net.Uri;

import java.security.InvalidParameterException;
import java.util.HashMap;

public class Post {
	public boolean isNew;
	public int post_id;
	public int road_id;
	public String comment;
	public int level;
	public String lang;
	public String post_time;
	public boolean is_steps;
	public boolean is_difference;
	public boolean is_steep;
	public boolean is_rough;
	public boolean is_narrow;
	public boolean is_cant;
	public boolean is_bikes;

	public Post( int road_id, String comment, int level, String lang,
	             HashMap< String, Boolean > types ) {
		this.isNew = true;

		this.road_id = road_id;
		this.comment = comment;
		this.level   = level;
		this.lang    = lang;
		is_steps      = types.get( "is_steps" );
		is_difference = types.get( "is_difference" );
		is_steep      = types.get( "is_steep" );
		is_rough      = types.get( "is_rough" );
		is_narrow     = types.get( "is_narrow" );
		is_cant       = types.get( "is_cant" );
		is_bikes      = types.get( "is_bikes" );
	}

	public Post( int post_id, String comment, int level, String lang, String post_time,
	             HashMap< String, Boolean > types ) {
		this.isNew = false;

		this.post_id = post_id;
		this.comment = comment;
		this.level   = level;
		this.lang    = lang;
		this.post_time    = post_time;
		is_steps      = types.get( "is_steps" );
		is_difference = types.get( "is_difference" );
		is_steep      = types.get( "is_steep" );
		is_rough      = types.get( "is_rough" );
		is_narrow     = types.get( "is_narrow" );
		is_cant       = types.get( "is_cant" );
		is_bikes      = types.get( "is_bikes" );
	}

	public String toString() {
		String res = "";

		switch ( level ) {
			case 0:
				res += "付き添いなしでも快適に通行可";
				break;

			case 1:
				res += "付き添いなしでは労力を要する\n";
				break;

			case 2:
				res += "付き添いが必要\n";
				break;

			case 3:
				res += "通行不能\n";
				break;
		}

		res += is_steps ? "階段 " : "";
		res += is_difference ? "段差 " : "";
		res += is_steep ? "急な坂 " : "";
		res += is_rough ? "路面が悪い " : "";
		res += is_narrow ? "狭い " : "";
		res += is_cant ? "横方向の傾斜 " : "";
		res += is_bikes ? "塞がれている " : "";
		res += "\n";

		res += comment + "\n";
		res += post_time;

		return res;
	}

	public void append( Uri.Builder builder ) {
		if ( ! isNew ) {
			throw new InvalidParameterException();
		}

		builder.appendQueryParameter( "road_id", String.valueOf( road_id ) );
		builder.appendQueryParameter( "comment", comment );
		builder.appendQueryParameter( "level", String.valueOf( level ) );
		builder.appendQueryParameter( "lang", lang );
		builder.appendQueryParameter( "is_steps", is_steps ? "1" : "0" );
		builder.appendQueryParameter( "is_difference", is_difference ? "1" : "0" );
		builder.appendQueryParameter( "is_steep", is_steep ? "1" : "0" );
		builder.appendQueryParameter( "is_rough", is_rough ? "1" : "0" );
		builder.appendQueryParameter( "is_narrow", is_narrow ? "1" : "0" );
		builder.appendQueryParameter( "is_cant", is_cant ? "1" : "0" );
		builder.appendQueryParameter( "is_bikes", is_bikes ? "1" : "0" );
	}
}
